compiler/p/codegen/OMRTreeEvaluator.cpp

```diff
@@ -5826,6 +5826,78 @@ TR::Register *OMR::Power::TreeEvaluator::retrieveTOCRegister(TR::Node *node, TR:
    return cg->machine()->getRealRegister(TR::RealRegister::gr2);
 }
 
+TR::Register *OMR::Power::TreeEvaluator::sbyteswapEvaluator(TR::Node *node, TR::CodeGenerator *cg)
+   {
+   TR_ASSERT(node->getNumChildren() == 1, "Wrong number of children in sbyteswapEvaluator");
+
+   TR::Node *firstChild = node->getFirstChild();
+   TR::Register *tgtRegister = cg->allocateRegister();
+
+   TR::Node *firstNonConversionOpCodeNode = node->getFirstChild();
+   TR::DataType nodeType = firstNonConversionOpCodeNode->getType();
+
+   //Move through descendants until a non conversion opcode is reached,
+   //while making sure all nodes have a ref count of 1 and the types are between 2-8 bytes
+   while (firstNonConversionOpCodeNode->getOpCode().isConversion() &&
+          firstNonConversionOpCodeNode->getReferenceCount() == 1 &&
+          (nodeType.isInt16() || nodeType.isInt32() || nodeType.isInt64()))
+      {
+      firstNonConversionOpCodeNode = firstNonConversionOpCodeNode->getFirstChild();
+      nodeType = firstNonConversionOpCodeNode->getType();
+      }
+
+   if (!firstNonConversionOpCodeNode->getRegister() &&
+       firstNonConversionOpCodeNode->getOpCode().isMemoryReference() &&
+       firstNonConversionOpCodeNode->getReferenceCount() == 1 &&
+       (nodeType.isInt16() || nodeType.isInt32() || nodeType.isInt64()))
+      {
+      TR::MemoryReference *tempMR = new (cg->trHeapMemory()) TR::MemoryReference(firstNonConversionOpCodeNode, 2, cg);
+#ifndef __LITTLE_ENDIAN__
+      //On Big Endian Machines
+      if (nodeType.isInt32())
+         tempMR->addToOffset(node,2,cg);
+      else if (nodeType.isInt64())
+         tempMR->addToOffset(node,6,cg);
+#endif
+      tempMR->forceIndexedForm(firstNonConversionOpCodeNode, cg);
+      generateTrg1MemInstruction(cg, TR::InstOpCode::lhbrx, node, tgtRegister, tempMR);
```

Comment:
This seems error prone. I don't suggest we change this in this PR of course, but we should consider as a separate work item to push forcing of index form into `generateTrg1MemInstruction` if this always needs to be performed. A similar thing happens on Z. See example in [1]. There are tons of others in that file. This way the user doesn't have to worry whether they need the memory reference to look a certain way when generating an instruction using the said memory reference.

[1] https://github.com/eclipse/omr/blob/16b345c029ac35b0478c42e18320378f6feacadc/compiler/z/codegen/S390GenerateInstructions.cpp#L1525-L1530

Reply:
Agreed. I've opened #5540 to track this.